### ISSUE TITLE
CNI plugin fails to configure pods in IPv6 clusters

### ISSUE THREAD

**Issue by emike922** [istio/istio#36871]
### Bug Description

When deploying workloads into the service mesh while running the Istio CNI plugin in an IPv6 kind cluster, the workload pods seem to always end up in a cyclic crash loop. Using istio-init (ie w/o CNI plugin) appears to works fine.
Am I missing something?

Is it expected that CNI reports ENABLE_INBOUND_IPV6=false (while istio-init in a similar environment reports true)?

**Edit:** I observed the same behavior in a native IPv6 cluster with docker/k8s hosts assigned only IPv6 addresses.

```
$ cat ipv6.yaml 
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  ipFamily: ipv6
$ kind create cluster --image=kindest/node:v1.21.2 --name=ipv6 --config=ipv6.yaml

$ helm upgrade istio-cni \
    manifests/charts/istio-cni \
    --install \
    --namespace kube-system \
    --set global.hub=gcr.io/istio-release \
    --set global.tag=1.12.1 \
    --set cni.logLevel=debug

$ helm upgrade istio-base \
    manifests/charts/base \
    --install \
    --namespace istio-system \
    --create-namespace

$ helm upgrade istiod \
    manifests/charts/istio-control/istio-discovery \
    --install \
    --namespace istio-system \
    --set global.hub=gcr.io/istio-release \
    --set global.tag=1.12.1 \
    --set istio_cni.enabled=true \
    --wait

$ kubectl create namespace test
$ kubectl label namespace test istio-injection=enabled
$ kubectl apply -n test -f samples/sleep/sleep.yaml

$ kubectl get pods -n test -w
NAME                     READY   STATUS     RESTARTS   AGE
sleep-557747455f-gt5f4   0/2     Init:0/1   0          14s
sleep-557747455f-gt5f4   0/2     Init:0/1   0          40s
sleep-557747455f-gt5f4   0/2     Init:Error   0          47s
sleep-557747455f-gt5f4   0/2     Init:0/1     1          51s
sleep-557747455f-gt5f4   0/2     Terminating   1          51s
sleep-557747455f-gt5f4   0/2     Terminating   1          52s
sleep-557747455f-rdtfk   0/2     Pending       0          0s
sleep-557747455f-rdtfk   0/2     Pending       0          0s
sleep-557747455f-rdtfk   0/2     Init:0/1      0          1s
sleep-557747455f-rdtfk   0/2     Init:0/1      0          13s
sleep-557747455f-rdtfk   0/2     Init:Error    0          19s
sleep-557747455f-rdtfk   0/2     Init:0/1      1          22s
sleep-557747455f-rdtfk   0/2     Terminating   1          22s
sleep-557747455f-rdtfk   0/2     Terminating   1          23s
sleep-557747455f-kg7tw   0/2     Pending       0          1s
sleep-557747455f-kg7tw   0/2     Pending       0          1s
sleep-557747455f-kg7tw   0/2     Init:0/1      0          2s
sleep-557747455f-kg7tw   0/2     Init:0/1      0          16s
sleep-557747455f-kg7tw   0/2     Init:Error    0          22s
...

$ kubectl logs -n kube-system istio-cni-node-phw5h
...
2022-01-18T15:25:37.564975Z     info    repair  Pod detected as broken, deleting: test/sleep-557747455f-rdtfk
2022-01-18T15:25:38.237229Z     info    repair  Pod detected as broken, deleting: test/sleep-557747455f-rdtfk
2022-01-18T15:25:41.916866Z     info    cni     istio-cni cmdAdd with k8s args: {CommonArgs:{IgnoreUnknown:true} IP:<nil> K8S_POD_NAME:sleep-557747455f-kg7tw K8S_POD_NAMESPACE:test K8S_POD_INFRA_CONTAINER_ID:12f1622e01ad06335c4cd7ae05bfb0939b40a7991729e6c96f2b24a79101d930}
2022-01-18T15:25:41.916888Z     info    cni     ============= Start iptables configuration for sleep-557747455f-kg7tw =============
2022-01-18T15:25:41.916892Z     info    cni     Istio iptables environment:
ENVOY_PORT=
INBOUND_CAPTURE_PORT=
ISTIO_INBOUND_INTERCEPTION_MODE=
ISTIO_INBOUND_TPROXY_ROUTE_TABLE=
ISTIO_INBOUND_PORTS=
ISTIO_OUTBOUND_PORTS=
ISTIO_LOCAL_EXCLUDE_PORTS=
ISTIO_EXCLUDE_INTERFACES=
ISTIO_SERVICE_CIDR=
ISTIO_SERVICE_EXCLUDE_CIDR=
ISTIO_META_DNS_CAPTURE=
2022-01-18T15:25:41.916894Z     info    cni     Istio iptables variables:
PROXY_PORT=15001
PROXY_INBOUND_CAPTURE_PORT=15006
PROXY_TUNNEL_PORT=15008
PROXY_UID=1337
PROXY_GID=1337
INBOUND_INTERCEPTION_MODE=REDIRECT
INBOUND_TPROXY_MARK=1337
INBOUND_TPROXY_ROUTE_TABLE=133
INBOUND_PORTS_INCLUDE=*
INBOUND_PORTS_EXCLUDE=15020,15021,15090
OUTBOUND_IP_RANGES_INCLUDE=*
OUTBOUND_IP_RANGES_EXCLUDE=
OUTBOUND_PORTS_INCLUDE=
OUTBOUND_PORTS_EXCLUDE=15020
KUBE_VIRT_INTERFACES=
ENABLE_INBOUND_IPV6=false
DNS_CAPTURE=false
CAPTURE_ALL_DNS=false
DNS_SERVERS=[],[]
OUTPUT_PATH=
NETWORK_NAMESPACE=/var/run/netns/cni-c6459cec-3d6e-cdb4-a14a-433b029d1043
CNI_MODE=true
EXCLUDE_INTERFACES=
2022-01-18T15:25:41.916897Z     info    cni     Writing following contents to rules file: /tmp/iptables-rules-1642519541608327696.txt3276312655
* nat
-N ISTIO_INBOUND
-N ISTIO_REDIRECT
-N ISTIO_IN_REDIRECT
-N ISTIO_OUTPUT
-A ISTIO_INBOUND -p tcp --dport 15008 -j RETURN
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A ISTIO_INBOUND -p tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15021 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_OUTPUT -p tcp --dport 15020 -j RETURN
-A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
COMMIT
2022-01-18T15:25:41.916899Z     info    cni     Running command: nsenter --net=/var/run/netns/cni-c6459cec-3d6e-cdb4-a14a-433b029d1043 -- iptables-restore --noflush /tmp/iptables-rules-1642519541608327696.txt3276312655
2022-01-18T15:25:41.916902Z     info    cni     Writing following contents to rules file: /tmp/ip6tables-rules-1642519541845170871.txt3226216103
2022-01-18T15:25:41.916904Z     info    cni     Running command: nsenter --net=/var/run/netns/cni-c6459cec-3d6e-cdb4-a14a-433b029d1043 -- ip6tables-restore --noflush /tmp/ip6tables-rules-1642519541845170871.txt3226216103
2022-01-18T15:25:41.916905Z     info    cni     Running command: nsenter --net=/var/run/netns/cni-c6459cec-3d6e-cdb4-a14a-433b029d1043 -- iptables-save
2022-01-18T15:25:41.916908Z     info    cni     Command output: 
# Generated by iptables-save v1.8.7 on Tue Jan 18 15:25:41 2022
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:ISTIO_INBOUND - [0:0]
:ISTIO_IN_REDIRECT - [0:0]
:ISTIO_OUTPUT - [0:0]
:ISTIO_REDIRECT - [0:0]
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_INBOUND -p tcp -m tcp --dport 15008 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15021 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A ISTIO_OUTPUT -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
COMMIT
# Completed on Tue Jan 18 15:25:41 2022
2022-01-18T15:25:41.916913Z     info    cni     Running command: nsenter --net=/var/run/netns/cni-c6459cec-3d6e-cdb4-a14a-433b029d1043 -- /opt/cni/bin/istio-cni configure-routes
2022-01-18T15:25:41.916916Z     error   cni     Command error output: 
{"level":"info","time":"2022-01-18T15:25:41.913060Z","msg":"Added ::6 address"}
2022-01-18T15:25:41.916917Z     info    cni     ============= End iptables configuration for sleep-557747455f-kg7tw =============
```

### Version

```prose
Istio 1.12.1
Kubernetes 1.21.2
```

**Comment by mKeRix** [comment]
I believe I stumbled upon this issue as well while trying to fix Istio for EKS on IPv6. For testing reasons I have now hardcoded that config value to always be true and can confirm that this works. Looking at the code my wild guess is that this will happen when you run pods on hosts that are dualstack, as Istio seems to prefer IPv4 in a lot of places. This would explain the difference in behavior for me at least:

- running this within istio-proxy will check the local IPs in the pod namespace, which are IPv6 only
- running this within the cni will execute the code on the host network namespace, which may have an IPv4 address attached

That being said I've only started looking at the code today, so my guess may also be way off base.

**Comment by mKeRix** [comment]
I believe I have a fix for this in my local code base, by switching to the pod network namespace before running the iptables code. I will test it a bit more and then open a PR.

**Comment by mKeRix** [comment]
Should be possible, however that would require a bigger change I believe. Currently the IP is determined in the same istio-iptables tool for both cni and regular setups. Since the prevresult is not available within a pod we'd have to provide a way to override the detection flow there and instead use a static list of IPs that are optionally passed as an argument.

In contrast to that, switching the network namespace only requires slight adaptions to the iptables part of the cni plugin - so it seemed more straightforward to me that way.

**Comment by lwj5** [comment]
> I believe I stumbled upon this issue as well while trying to fix Istio for EKS on IPv6. For testing reasons I have now hardcoded that config value to always be true and can confirm that this works. Looking at the code my wild guess is that this will happen when you run pods on hosts that are dualstack, as Istio seems to prefer IPv4 in a lot of places. This would explain the difference in behavior for me at least:
> 
> * running this within istio-proxy will check the local IPs in the pod namespace, which are IPv6 only
> * running this within the cni will execute the code on the host network namespace, which may have an IPv4 address attached
> 
> That being said I've only started looking at the code today, so my guess may also be way off base.

may i know how you changed the config value to true?

I'm running on EKS cluster with the cni deployed using Helm, and my pods have the same cycle of crash loop

**Comment by kcm-krishna-git** [comment]
> > I believe I stumbled upon this issue as well while trying to fix Istio for EKS on IPv6. For testing reasons I have now hardcoded that config value to always be true and can confirm that this works. Looking at the code my wild guess is that this will happen when you run pods on hosts that are dualstack, as Istio seems to prefer IPv4 in a lot of places. This would explain the difference in behavior for me at least:
> > 
> > * running this within istio-proxy will check the local IPs in the pod namespace, which are IPv6 only
> > * running this within the cni will execute the code on the host network namespace, which may have an IPv4 address attached
> > 
> > That being said I've only started looking at the code today, so my guess may also be way off base.
> 
> may i know how you changed the config value to true?
> 
> I'm running on IPv6 EKS cluster with the cni deployed using Helm 1.20.1, and my pods have the same cycle of crash loop
I am facing the similar issue with istio 1.20.4 on RKE2. Is the fix mentioned above to work with dual stack OS merged to 1.20.4?
